File: swift-sdks/Sources/SuiSwiftSDK/DeepBookV3/DeepBookAdmin.swift
import Foundation

// Admin Operations for Swift SDK (P3)
public class DeepBookAdmin {
    weak var client: DeepBookV3Client?

    public init(client: DeepBookV3Client) {
        self.client = client
    }

    // MARK: - Admin Operations

    public func createPoolAdmin(
        sender: String,
        baseCoinKey: String,
        quoteCoinKey: String,
        tickSize: Double,
        lotSize: Double,
        minSize: Double
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: "\((client?.config.deepbookPackage ?? ""))::pool::create_pool_admin",
            arguments: [
                [
                    "baseCoinKey": baseCoinKey,
                    "quoteCoinKey": quoteCoinKey,
                    "tickSize": tickSize,
                    "lotSize": lotSize,
                    "minSize": minSize
                ]
            ]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func setTickSize(
        sender: String,
        poolKey: String,
        tickSize: Double
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: "\((client?.config.deepbookPackage ?? ""))::pool::set_tick_size",
            arguments: [["poolKey": poolKey, "tickSize": tickSize]]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func setLotSize(
        sender: String,
        poolKey: String,
        lotSize: Double
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: "\((client?.config.deepbookPackage ?? ""))::pool::set_lot_size",
            arguments: [["poolKey": poolKey, "lotSize": lotSize]]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func setBasePricePoint(
        sender: String,
        poolKey: String,
        price: UInt64
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: "\((client?.config.deepbookPackage ?? ""))::pool::set_base_price_point",
            arguments: [["poolKey": poolKey, "price": price]]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func setQuotePricePoint(
        sender: String,
        poolKey: String,
        price: UInt64
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: "\((client?.config.deepbookPackage ?? ""))::pool::set_quote_price_point",
            arguments: [["poolKey": poolKey, "price": price]]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func updatePoolAllowedVersions(
        sender: String,
        poolKey: String,
        registryId: String
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: "\((client?.config.deepbookPackage ?? ""))::pool::update_pool_allowed_versions",
            arguments: [["poolKey": poolKey, "registryId": registryId]]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func withdrawAll(
        sender: String,
        poolKey: String,
        recipient: String
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: "\((client?.config.deepbookPackage ?? ""))::pool::withdraw_all",
            arguments: [["poolKey": poolKey, "recipient": recipient]]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func removePool(
        sender: String,
        poolKey: String
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: "\((client?.config.deepbookPackage ?? ""))::pool::remove_pool",
            arguments: [["poolKey": poolKey]]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func burnDeep(
        sender: String,
        poolKey: String,
        treasuryId: String
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: "\((client?.config.deepbookPackage ?? ""))::pool::burn_deep",
            arguments: [["poolKey": poolKey, "treasuryId": treasuryId]]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    // MARK: - Helper Methods

    private func call(_ method: String, params params: [Any?] = []) async throws -> [String: Any] {
        guard let client = client else {
            throw DeepBookError.invalidResponse
        }
        return try await client.call(method, params: params)
    }

    private func buildTransaction(target: String, arguments: [[String: Any]]) -> [String: Any] {
        var transactions: [[String: Any]] = []
        
        for args in arguments {
            var moveCall: [String: Any] = [:]
            moveCall["kind"] = "moveCall"
            moveCall["target"] = target
            moveCall["arguments"] = args
            transactions.append(moveCall)
        }

        return [
            "kind": "programmableTransaction",
            "inputs": [],
            "transactions": transactions
        ]
    }
}

File: java-sdks/src/main/java/com/suisdks/sui/deepbook_v3/DeepBookAdmin.java
package com.suisdks.sui.deepbook_v3;

import java.math.BigInteger;
import java.util.*;

public class DeepBookAdmin {
    private final DeepBookV3Client client;

    public DeepBookAdmin(DeepBookV3Client client) {
        this.client = client;
    }

    public DeepBookV3Client client() {
        return client;
    }

    // ==================== Admin Operations ====================

    public Map<String, Object> createPoolAdmin(String sender, String baseCoinKey, String quoteCoinKey,
            double tickSize, double lotSize, double minSize) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");
        
        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().deepbookPackage() + "::pool::create_pool_admin");
        
        Map<String, Object> args = new HashMap<>();
        args.put("baseCoinKey", baseCoinKey);
        args.put("quoteCoinKey", quoteCoinKey);
        args.put("tickSize", tickSize);
        args.put("lotSize", lotSize);
        args.put("minSize", minSize);
        moveCall.put("arguments", List.of(args));
        
        txData.put("inputs", List.of());
        txData.put("transactions", List.of(moveCall));

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Map<String, Object> setTickSize(String sender, String poolKey, double tickSize) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");
        
        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().deepbookPackage() + "::pool::set_tick_size");
        
        Map<String, Object> args = new HashMap<>();
        args.put("poolKey", poolKey);
        args.put("tickSize", tickSize);
        moveCall.put("arguments", List.of(args));
        
        txData.put("inputs", List.of());
        txData.put("transactions", List.of(moveCall));

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Map<String, Object> setLotSize(String sender, String poolKey, double lotSize) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");
        
        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().deepbookPackage() + "::pool::set_lot_size");
        
        Map<String, Object> args = new HashMap<>();
        args.put("poolKey", poolKey);
        args.put("lotSize", lotSize);
        moveCall.put("arguments", List.of(args));
        
        txData.put("inputs", List.of());
        txData.put("transactions", List.of(moveCall));

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Map<String, Object> setBasePricePoint(String sender, String poolKey, BigInteger price) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");
        
        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().deepbookPackage() + "::pool::set_base_price_point");
        
        Map<String, Object> args = new HashMap<>();
        args.put("poolKey", poolKey);
        args.put("price", price.toString());
        moveCall.put("arguments", List.of(args));
        
        txData.put("inputs", List.of());
        txData.put("transactions", List.of(moveCall));

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Map<String, Object> setQuotePricePoint(String sender, String poolKey, BigInteger price) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");
        
        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().deepbookPackage() + "::pool::set_quote_price_point");
        
        Map<String, Object> args = new HashMap<>();
        args.put("poolKey", poolKey);
        args.put("price", price.toString());
        moveCall.put("arguments", List.of(args));
        
        txData.put("inputs", List.of());
        txData.put("transactions", List.of(moveCall));

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Map<String, Object> updatePoolAllowedVersions(String sender, String poolKey, String registryId) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");
        
        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().deepbookPackage() + "::pool::update_pool_allowed_versions");
        
        Map<String, Object> args = new HashMap<>();
        args.put("poolKey", poolKey);
        args.put("registryId", registryId);
        moveCall.put("arguments", List.of(args));
        
        txData.put("inputs", List.of());
        txData.put("transactions", List.of(moveCall));

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Map<String, Object> withdrawAll(String sender, String poolKey, String recipient) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");
        
        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().deepbookPackage() + "::pool::withdraw_all");
        
        Map<String, Object> args = new HashMap<>();
        args.put("poolKey", poolKey);
        args.put("recipient", recipient);
        moveCall.put("arguments", List.of(args));
        
        txData.put("inputs", List.of());
        txData.put("transactions", List.of(moveCall));

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Map<String, Object> removePool(String sender, String poolKey) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");
        
        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().deepbookPackage() + "::pool::remove_pool");
        
        Map<String, Object> args = new HashMap<>();
        args.put("poolKey", poolKey);
        moveCall.put("arguments", List.of(args));
        
        txData.put("inputs", List.of());
        txData.put("transactions", List.of(moveCall));

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Map<String, Object> burnDeep(String sender, String poolKey, String treasuryId) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");
        
        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().deepbookPackage() + "::pool::burn_deep");
        
        Map<String, Object> args = new HashMap<>();
        args.put("poolKey", poolKey);
        args.put("treasuryId", treasuryId);
        moveCall.put("arguments", List.of(args));
        
        txData.put("inputs", List.of());
        txData.put("transactions", List.of(moveCall));

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }
}